File: src/main/ioke/lang/RegexpMatch.java
/*
 * See LICENSE file in distribution for copyright and licensing information.
 */
package ioke.lang;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.Set;

import org.jregex.MatchResult;

import ioke.lang.exceptions.ControlFlow;

/**
 * @author <a href="mailto:ola.bini@gmail.com">Ola Bini</a>
 */
public class RegexpMatch extends IokeData {
    private IokeObject regexp;
    private MatchResult mr;
    private IokeObject target;

    public RegexpMatch(IokeObject regexp, MatchResult mr, IokeObject target) {
        this.regexp = regexp;
        this.mr = mr;
        this.target = target;
    }
    
    public static Object getTarget(Object on) throws ControlFlow {
        return ((RegexpMatch)IokeObject.data(on)).target;
    }

    public static Object getRegexp(Object on) throws ControlFlow {
        return ((RegexpMatch)IokeObject.data(on)).regexp;
    }

    public static MatchResult getMatchResult(Object on) throws ControlFlow {
        return ((RegexpMatch)IokeObject.data(on)).mr;
    }

    @Override
    public void init(final IokeObject obj) throws ControlFlow {
        final Runtime runtime = obj.runtime;
        obj.setKind("Regexp Match");

        obj.registerMethod(runtime.newJavaMethod("Returns the target that this match was created against", new TypeCheckingJavaMethod.WithNoArguments("target", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return getTarget(on);
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("returns a list of all the named groups in the regular expression used to create this match", new TypeCheckingJavaMethod.WithNoArguments("names", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    Set names = Regexp.getRegexp(getRegexp(on)).getGroupNames();
                    List<Object> theNames = new ArrayList<Object>();
                    for(Object name : names) {
                        theNames.add(context.runtime.getSymbol(((String)name)));
                    }
                    return context.runtime.newList(theNames);
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("returns the part of the target before the text that matched", new TypeCheckingJavaMethod.WithNoArguments("beforeMatch", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return context.runtime.newText(getMatchResult(on).prefix());
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("returns the part of the target after the text that matched", new TypeCheckingJavaMethod.WithNoArguments("afterMatch", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return context.runtime.newText(getMatchResult(on).suffix());
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("returns the text that matched", new TypeCheckingJavaMethod.WithNoArguments("match", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return context.runtime.newText(getMatchResult(on).group(0));
                }
            }));

        obj.aliasMethod("match", "asText", null, null);

        obj.registerMethod(runtime.newJavaMethod("returns the number of groups available in this match", new TypeCheckingJavaMethod.WithNoArguments("length", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return context.runtime.newNumber(getMatchResult(on).groupCount());
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("returns a list of all groups captured in this match. if a group is not matched it will be nil in the list. the actual match text is not included in this list.", new TypeCheckingJavaMethod.WithNoArguments("captures", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    List<Object> groups = new ArrayList<Object>();
                    MatchResult mr = getMatchResult(on);
                    int len = mr.groupCount();
                    for(int i=1;i<len;i++) {
                        if(mr.isCaptured(i)) {
                            groups.add(context.runtime.newText(mr.group(i)));
                        } else {
                            groups.add(context.runtime.nil);
                        }
                    }

                    return context.runtime.newList(groups);
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("returns a list of all groups captured in this match. if a group is not matched it will be nil in the list. the actual match text is the first element in the list.", new TypeCheckingJavaMethod.WithNoArguments("asList", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    List<Object> groups = new ArrayList<Object>();
                    MatchResult mr = getMatchResult(on);
                    int len = mr.groupCount();
                    for(int i=0;i<len;i++) {
                        if(mr.isCaptured(i)) {
                            groups.add(context.runtime.newText(mr.group(i)));
                        } else {
                            groups.add(context.runtime.nil);
                        }
                    }

                    return context.runtime.newList(groups);
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one optional argument that should be either a number or a symbol. this should be the name or index of a group to return the start index for. if no index is supplied, 0 is the default. if the group in question wasn't matched, returns -1.", new TypeCheckingJavaMethod("start") {
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .receiverMustMimic(obj)
                    .withOptionalPositional("index", "0")
                    .getArguments();

                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    int index = 0;
                    
                    if(args.size() > 0) {
                        Object arg = args.get(0);
                        if(IokeObject.data(arg) instanceof Number) {
                            index = Number.extractInt(arg, message, context);
                        } else {
                            String namedIndex = Text.getText(context.runtime.asText.sendTo(context, arg));
                            Integer ix = Regexp.getRegexp(getRegexp(on)).groupId(namedIndex);
                            if(ix == null) {
                                return context.runtime.newNumber(-1);
                            }
                            
                            index = ix;
                        }
                    }
                    MatchResult mr = getMatchResult(on);
                    if(index < mr.groupCount() && mr.isCaptured(index)) {
                        return context.runtime.newNumber(mr.start(index));
                    } else {
                        return context.runtime.newNumber(-1);
                    }
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one optional argument that should be either a number or a symbol. this should be the name or index of a group to return the end index for. if no index is supplied, 0 is the default. if the group in question wasn't matched, returns -1.", new TypeCheckingJavaMethod("end") {
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .receiverMustMimic(obj)
                    .withOptionalPositional("index", "0")
                    .getArguments();

                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    int index = 0;
                    
                    if(args.size() > 0) {
                        Object arg = args.get(0);
                        if(IokeObject.data(arg) instanceof Number) {
                            index = Number.extractInt(arg, message, context);
                        } else {
                            String namedIndex = Text.getText(context.runtime.asText.sendTo(context, arg));
                            Integer ix = Regexp.getRegexp(getRegexp(on)).groupId(namedIndex);
                            if(ix == null) {
                                return context.runtime.newNumber(-1);
                            }
                            
                            index = ix;
                        }
                    }
                    MatchResult mr = getMatchResult(on);
                    if(index < mr.groupCount() && mr.isCaptured(index)) {
                        return context.runtime.newNumber(mr.end(index));
                    } else {
                        return context.runtime.newNumber(-1);
                    }
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one optional argument that should be either a number or a symbol. this should be the name or index of a group to return the start and end index for. if no index is supplied, 0 is the default. if the group in question wasn't matched, returns nil, otherwise a pair of the start and end indices.", new TypeCheckingJavaMethod("offset") {
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .receiverMustMimic(obj)
                    .withOptionalPositional("index", "0")
                    .getArguments();

                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    int index = 0;
                    
                    if(args.size() > 0) {
                        Object arg = args.get(0);
                        if(IokeObject.data(arg) instanceof Number) {
                            index = Number.extractInt(arg, message, context);
                        } else {
                            String namedIndex = Text.getText(context.runtime.asText.sendTo(context, arg));
                            Integer ix = Regexp.getRegexp(getRegexp(on)).groupId(namedIndex);
                            if(ix == null) {
                                return context.runtime.nil;
                            }
                            
                            index = ix;
                        }
                    }
                    MatchResult mr = getMatchResult(on);
                    if(index < mr.groupCount() && mr.isCaptured(index)) {
                        return context.runtime.newPair(context.runtime.newNumber(mr.start(index)), context.runtime.newNumber(mr.end(index)));
                    } else {
                        return context.runtime.nil;
                    }
                }
            }));


        obj.registerMethod(runtime.newJavaMethod("Takes one indexing argument that should be either a number, a range, a text or a symbol. if it's a number or a range of numbers, these will specify the index of the capture to return. 0 is the whole match. negative indices are interpreted in the usual way. if the range is out of range it will only use as many groups as there are. if it's a text or a sym it will be interpreted as a the name of a named group to return. if an index isn't correct or wasn't matched, it returns nil in those places.", new TypeCheckingJavaMethod("[]") {
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .receiverMustMimic(obj)
                    .withRequiredPositional("index")
                    .getArguments();

                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    Object arg = args.get(0);

                    MatchResult mr = getMatchResult(on);

                    if((IokeObject.data(arg) instanceof Symbol) || (IokeObject.data(arg) instanceof Text)) {
                        String namedIndex = Text.getText(context.runtime.asText.sendTo(context, arg));
                        Integer ix = Regexp.getRegexp(getRegexp(on)).groupId(namedIndex);
                        if(ix == null || !mr.isCaptured(ix)) {
                            return context.runtime.nil;
                        }
                        return context.runtime.newText(mr.group(ix));
                    } else {
                        int size = mr.groupCount();

                        if(IokeObject.data(arg) instanceof Range) {
                            int first = Number.extractInt(Range.getFrom(arg), message, context); 
                        
                            if(first < 0) {
                                return context.runtime.newList(new ArrayList<Object>());
                            }

                            int last = Number.extractInt(Range.getTo(arg), message, context);
                            boolean inclusive = Range.isInclusive(arg);


                            if(last < 0) {
                                last = size + last;
                            }

                            if(last < 0) {
                                return context.runtime.newList(new ArrayList<Object>());
                            }

                            if(last >= size) {
                                last = inclusive ? size-1 : size;
                            }

                            if(first > last || (!inclusive && first == last)) {
                                return context.runtime.newList(new ArrayList<Object>());
                            }
                        
                            if(!inclusive) {
                                last--;
                            }
                        
                            List<Object> result = new ArrayList<Object>();
                            for(int i = first; i < last+1; i++) {
                                if(!mr.isCaptured(i)) {
                                    result.add(context.runtime.nil);
                                } else {
                                    result.add(context.runtime.newText(mr.group(i)));
                                }
                            }

                            return context.runtime.newList(result);
                        }
                        
                        if(!(IokeObject.data(arg) instanceof Number)) {
                            arg = IokeObject.convertToNumber(arg, message, context);
                        }
                        int index = ((Number)IokeObject.data(arg)).asJavaInteger();
                        if(index < 0) {
                            index = size + index;
                        }

                        if(index >= 0 && index < size && mr.isCaptured(index)) {
                            return context.runtime.newText(mr.group(index));
                        } else {
                            return context.runtime.nil;
                        }
                    }
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("will get the named group corresponding to the name of the message, or nil if the named group hasn't been matched. will signal a condition if no such group is defined.", new TypeCheckingJavaMethod("pass") {
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .receiverMustMimic(obj)
                    .getArguments();
                
                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }
                
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    MatchResult mr = getMatchResult(on);
                    String name = Message.name(message);
                    
                    Integer ix = Regexp.getRegexp(getRegexp(on)).groupId(name);
                    if(ix == null) {
                        final IokeObject condition = IokeObject.as(IokeObject.getCellChain(message.runtime.condition, 
                                                                                           message, 
                                                                                           context, 
                                                                                           "Error", 
                                                                                           "NoSuchCell"), context).mimic(message, context);
                        condition.setCell("message", message);
                        condition.setCell("context", context);
                        condition.setCell("receiver", on);
                        condition.setCell("cellName", message.runtime.getSymbol(name));

                        message.runtime.withReturningRestart("ignore", context, new RunnableWithControlFlow() {
                                public void run() throws ControlFlow {
                                    condition.runtime.errorCondition(condition);
                                }});

                        return context.runtime.nil;
                    }

                    if(mr.isCaptured(ix)) {
                        return context.runtime.newText(mr.group(ix));
                    } else {
                        return context.runtime.nil;
                    }
                }
            }));
        
    }
}

File: src/ikc/main/Ioke.Lang/RegexpMatch.cs

namespace Ioke.Lang {
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using NRegex;

    using Ioke.Lang.Util;

    public class RegexpMatch : IokeData {
        IokeObject regexp;
        MatchResult mr;
        IokeObject target;

        public RegexpMatch(IokeObject regexp, MatchResult mr, IokeObject target) {
            this.regexp = regexp;
            this.mr = mr;
            this.target = target;
        }
    
        public static object GetTarget(object on) {
            return ((RegexpMatch)IokeObject.dataOf(on)).target;
        }

        public static object GetRegexp(object on) {
            return ((RegexpMatch)IokeObject.dataOf(on)).regexp;
        }

        public static MatchResult GetMatchResult(object on) {
            return ((RegexpMatch)IokeObject.dataOf(on)).mr;
        }

        public override void Init(IokeObject obj) {
            Runtime runtime = obj.runtime;
            obj.Kind = "Regexp Match";

            obj.RegisterMethod(runtime.NewNativeMethod("Returns the target that this match was created against", 
                                                       new TypeCheckingNativeMethod.WithNoArguments("target", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return GetTarget(on);
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("returns a list of all the named groups in the regular expression used to create this match", 
                                                       new TypeCheckingNativeMethod.WithNoArguments("names", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        var names = Regexp.GetRegexp(GetRegexp(on)).GroupNames;
                                                                                                        var theNames = new SaneArrayList();
                                                                                                        foreach(object name in names) {
                                                                                                            theNames.Add(context.runtime.GetSymbol(((string)name)));
                                                                                                        }
                                                                                                        return context.runtime.NewList(theNames);
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("returns the part of the target before the text that matched", 
                                                       new TypeCheckingNativeMethod.WithNoArguments("beforeMatch", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return context.runtime.NewText(GetMatchResult(on).Prefix);
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("returns the part of the target after the text that matched", 
                                                       new TypeCheckingNativeMethod.WithNoArguments("afterMatch", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return context.runtime.NewText(GetMatchResult(on).Suffix);
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("returns the text that matched", 
                                                       new TypeCheckingNativeMethod.WithNoArguments("match", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return context.runtime.NewText(GetMatchResult(on).Group(0));
                                                                                                    })));

            obj.AliasMethod("match", "asText", null, null);

            obj.RegisterMethod(runtime.NewNativeMethod("returns the number of groups available in this match", 
                                                       new TypeCheckingNativeMethod.WithNoArguments("length", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return context.runtime.NewNumber(GetMatchResult(on).GroupCount);
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("returns a list of all groups captured in this match. if a group is not matched it will be nil in the list. the actual match text is not included in this list.", 
                                                       new TypeCheckingNativeMethod.WithNoArguments("captures", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        var groups = new SaneArrayList();
                                                                                                        MatchResult mr = GetMatchResult(on);
                                                                                                        int len = mr.GroupCount;
                                                                                                        for(int i=1;i<len;i++) {
                                                                                                            if(mr.IsCaptured(i)) {
                                                                                                                groups.Add(context.runtime.NewText(mr.Group(i)));
                                                                                                            } else {
                                                                                                                groups.Add(context.runtime.nil);
                                                                                                            }
                                                                                                        }

                                                                                                        return context.runtime.NewList(groups);
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("returns a list of all groups captured in this match. if a group is not matched it will be nil in the list. the actual match text is the first element in the list.", 
                                                       new TypeCheckingNativeMethod.WithNoArguments("asList", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        var groups = new SaneArrayList();
                                                                                                        MatchResult mr = GetMatchResult(on);
                                                                                                        int len = mr.GroupCount;
                                                                                                        for(int i=0;i<len;i++) {
                                                                                                            if(mr.IsCaptured(i)) {
                                                                                                                groups.Add(context.runtime.NewText(mr.Group(i)));
                                                                                                            } else {
                                                                                                                groups.Add(context.runtime.nil);
                                                                                                            }
                                                                                                        }

                                                                                                        return context.runtime.NewList(groups);
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one optional argument that should be either a number or a symbol. this should be the name or index of a group to return the start index for. if no index is supplied, 0 is the default. if the group in question wasn't matched, returns -1.", 
                                                       new TypeCheckingNativeMethod("start", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(obj)
                                                                                    .WithOptionalPositional("index", "0")
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        int index = 0;
                    
                                                                                        if(args.Count > 0) {
                                                                                            object arg = args[0];
                                                                                            if(IokeObject.dataOf(arg) is Number) {
                                                                                                index = Number.ExtractInt(arg, message, context);
                                                                                            } else {
                                                                                                string namedIndex = Text.GetText(((Message)IokeObject.dataOf(context.runtime.asText)).SendTo(context.runtime.asText, context, arg));
                                                                                                int ix = -1;
                                                                                                try {
                                                                                                    ix = Regexp.GetRegexp(GetRegexp(on)).GroupId(namedIndex);
                                                                                                } catch(Exception) {
                                                                                                    return context.runtime.NewNumber(-1);
                                                                                                }
                                                                                                index = ix;
                                                                                            }
                                                                                        }
                                                                                        MatchResult mr = GetMatchResult(on);
                                                                                        if(index < mr.GroupCount && mr.IsCaptured(index)) {
                                                                                            return context.runtime.NewNumber(mr.GetStart(index));
                                                                                        } else {
                                                                                            return context.runtime.NewNumber(-1);
                                                                                        }
                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one optional argument that should be either a number or a symbol. this should be the name or index of a group to return the end index for. if no index is supplied, 0 is the default. if the group in question wasn't matched, returns -1.", 
                                                       new TypeCheckingNativeMethod("end", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(obj)
                                                                                    .WithOptionalPositional("index", "0")
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        int index = 0;
                    
                                                                                        if(args.Count > 0) {
                                                                                            object arg = args[0];
                                                                                            if(IokeObject.dataOf(arg) is Number) {
                                                                                                index = Number.ExtractInt(arg, message, context);
                                                                                            } else {
                                                                                                string namedIndex = Text.GetText(((Message)IokeObject.dataOf(context.runtime.asText)).SendTo(context.runtime.asText, context, arg));
                                                                                                int ix = -1;
                                                                                                try {
                                                                                                    ix = Regexp.GetRegexp(GetRegexp(on)).GroupId(namedIndex);
                                                                                                } catch(Exception) {
                                                                                                    return context.runtime.NewNumber(-1);
                                                                                                }
                                                                                                index = ix;
                                                                                            }
                                                                                        }
                                                                                        MatchResult mr = GetMatchResult(on);
                                                                                        if(index < mr.GroupCount && mr.IsCaptured(index)) {
                                                                                            return context.runtime.NewNumber(mr.GetEnd(index));
                                                                                        } else {
                                                                                            return context.runtime.NewNumber(-1);
                                                                                        }
                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one optional argument that should be either a number or a symbol. this should be the name or index of a group to return the start and end index for. if no index is supplied, 0 is the default. if the group in question wasn't matched, returns nil, otherwise a pair of the start and end indices.", 
                                                       new TypeCheckingNativeMethod("offset", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(obj)
                                                                                    .WithOptionalPositional("index", "0")
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        int index = 0;
                    
                                                                                        if(args.Count > 0) {
                                                                                            object arg = args[0];
                                                                                            if(IokeObject.dataOf(arg) is Number) {
                                                                                                index = Number.ExtractInt(arg, message, context);
                                                                                            } else {
                                                                                                string namedIndex = Text.GetText(((Message)IokeObject.dataOf(context.runtime.asText)).SendTo(context.runtime.asText, context, arg));
                                                                                                int ix = -1;
                                                                                                try {
                                                                                                    ix = Regexp.GetRegexp(GetRegexp(on)).GroupId(namedIndex);
                                                                                                } catch(Exception) {
                                                                                                    return context.runtime.nil;
                                                                                                }
                                                                                                index = ix;
                                                                                            }
                                                                                        }
                                                                                        MatchResult mr = GetMatchResult(on);
                                                                                        if(index < mr.GroupCount && mr.IsCaptured(index)) {
                                                                                            return context.runtime.NewPair(context.runtime.NewNumber(mr.GetStart(index)), context.runtime.NewNumber(mr.GetEnd(index)));
                                                                                        } else {
                                                                                            return context.runtime.nil;
                                                                                        }
                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one indexing argument that should be either a number, a range, a text or a symbol. if it's a number or a range of numbers, these will specify the index of the capture to return. 0 is the whole match. negative indices are interpreted in the usual way. if the range is out of range it will only use as many groups as there are. if it's a text or a sym it will be interpreted as a the name of a named group to return. if an index isn't correct or wasn't matched, it returns nil in those places.", 
                                                       new TypeCheckingNativeMethod("[]", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(obj)
                                                                                    .WithRequiredPositional("index")
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        object arg = args[0];

                                                                                        MatchResult mr = GetMatchResult(on);

                                                                                        if((IokeObject.dataOf(arg) is Symbol) || (IokeObject.dataOf(arg) is Text)) {
                                                                                            string namedIndex = Text.GetText(((Message)IokeObject.dataOf(context.runtime.asText)).SendTo(context.runtime.asText, context, arg));
                                                                                            int ix = -1;
                                                                                            try {
                                                                                                ix = Regexp.GetRegexp(GetRegexp(on)).GroupId(namedIndex);
                                                                                            } catch(Exception) {
                                                                                                return context.runtime.nil;
                                                                                            }
                                                                                            if(!mr.IsCaptured(ix)) {
                                                                                                return context.runtime.nil;
                                                                                            }
                                                                                            return context.runtime.NewText(mr.Group(ix));
                                                                                        } else {
                                                                                            int size = mr.GroupCount;

                                                                                            if(IokeObject.dataOf(arg) is Range) {
                                                                                                int first = Number.ExtractInt(Range.GetFrom(arg), message, context); 
                        
                                                                                                if(first < 0) {
                                                                                                    return context.runtime.NewList(new SaneArrayList());
                                                                                                }

                                                                                                int last = Number.ExtractInt(Range.GetTo(arg), message, context);
                                                                                                bool inclusive = Range.IsInclusive(arg);


                                                                                                if(last < 0) {
                                                                                                    last = size + last;
                                                                                                }

                                                                                                if(last < 0) {
                                                                                                    return context.runtime.NewList(new SaneArrayList());
                                                                                                }

                                                                                                if(last >= size) {
                                                                                                    last = inclusive ? size-1 : size;
                                                                                                }

                                                                                                if(first > last || (!inclusive && first == last)) {
                                                                                                    return context.runtime.NewList(new SaneArrayList());
                                                                                                }
                        
                                                                                                if(!inclusive) {
                                                                                                    last--;
                                                                                                }
                        
                                                                                                var result = new SaneArrayList();
                                                                                                for(int i = first; i < last+1; i++) {
                                                                                                    if(!mr.IsCaptured(i)) {
                                                                                                        result.Add(context.runtime.nil);
                                                                                                    } else {
                                                                                                        result.Add(context.runtime.NewText(mr.Group(i)));
                                                                                                    }
                                                                                                }

                                                                                                return context.runtime.NewList(result);
                                                                                            }
                        
                                                                                            if(!(IokeObject.dataOf(arg) is Number)) {
                                                                                                arg = IokeObject.ConvertToNumber(arg, message, context);
                                                                                            }
                                                                                            int index = ((Number)IokeObject.dataOf(arg)).AsNativeInteger();
                                                                                            if(index < 0) {
                                                                                                index = size + index;
                                                                                            }

                                                                                            if(index >= 0 && index < size && mr.IsCaptured(index)) {
                                                                                                return context.runtime.NewText(mr.Group(index));
                                                                                            } else {
                                                                                                return context.runtime.nil;
                                                                                            }
                                                                                        }
                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("will get the named group corresponding to the name of the message, or nil if the named group hasn't been matched. will signal a condition if no such group is defined.", 
                                                       new TypeCheckingNativeMethod("pass", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(obj)
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        MatchResult mr = GetMatchResult(on);
                                                                                        string name = Message.GetName(message);
                    
                                                                                        int ix = -1;
                                                                                        try {
                                                                                            ix = Regexp.GetRegexp(GetRegexp(on)).GroupId(name);
                                                                                        } catch(Exception) {
                                                                                            IokeObject condition = IokeObject.As(IokeObject.GetCellChain(message.runtime.Condition, 
                                                                                                                                                         message, 
                                                                                                                                                         context, 
                                                                                                                                                         "Error", 
                                                                                                                                                         "NoSuchCell"), context).Mimic(message, context);
                                                                                            condition.SetCell("message", message);
                                                                                            condition.SetCell("context", context);
                                                                                            condition.SetCell("receiver", on);
                                                                                            condition.SetCell("cellName", message.runtime.GetSymbol(name));

                                                                                            message.runtime.WithReturningRestart("ignore", context, ()=>{condition.runtime.ErrorCondition(condition);});
                                                                                            return context.runtime.nil;
                                                                                        }

                                                                                        if(mr.IsCaptured(ix)) {
                                                                                            return context.runtime.NewText(mr.Group(ix));
                                                                                        } else {
                                                                                            return context.runtime.nil;
                                                                                        }
                                                                                    })));
        }
    }
}
